pkg/s3repository/repository_protected_entity.go

```diff
@@ -141,37 +197,304 @@ func (this ProtectedEntity) GetID() astrolabe.ProtectedEntityID {
+				segmentNumber++
+				uploadSegment = false
+				bytesThisSegment, err = skipBytes(reader, existingSegments[segmentNumber].length, nil)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		if uploadSegment {
+			bytesThisSegment, err = this.uploadSegment(ctx, name, partNum, startOffset, maxSegmentSize, reader)
+			if err != nil {
+				return err
+			}
+			if bytesThisSegment < maxSegmentSize {
+				return nil
+			}
+		}
+		partNum++
+		startOffset += bytesThisSegment
+	}
+	return nil
+}
 
-	result, err := uploader.Upload(&s3manager.UploadInput{
-		Body:   reader,
-		Bucket: aws.String(this.rpetm.bucket),
-		Key:    aws.String(name),
-	})
-	if err == nil {
-		log.Println("Successfully uploaded to", result.Location)
+/*
+ * skipBytes will either use the Seek API, if the Reader implements Seeker, or read and discard bytes from the
+ * Reader to move the offset forward.  If discardBuf is set, it will be used to read bytes into for discarding (this is
+ * simply to avoid  allocating a buffer when one is available)
+ */
+func skipBytes(reader io.Reader, bytesToSkip int64, discardBuf []byte) (int64, error) {
+	if bytesToSkip < 0 {
+		return 0, errors.New("Cannot skip negative bytes")
 	}
-	return err
+	if bytesToSkip == 0 {
+		return 0, nil	// Bounce out here so we'll never get down to allocating a discard buf
+	}
+	var bytesSkipped int64
+	if rs, ok := interface{}(reader).(io.Seeker); ok {
+		startOffset, err := rs.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return 0, nil
+		}
+
+		endOffset, err := rs.Seek(bytesToSkip, io.SeekCurrent)
+		if err != nil {
+			return 0, nil
+		}
+		bytesSkipped = endOffset - startOffset
+	} else {
+		if discardBuf == nil {
+			bufSize := bytesToSkip
+			if bytesToSkip > kMaxBufferSize {
+				bufSize = kMaxBufferSize
+			}
+			discardBuf = make([]byte, bufSize)
+		}
+		var bytesSkipped int64
+		for bytesSkipped < bytesToSkip {
+			curBuf := discardBuf
+			if bytesToSkip < int64(len(discardBuf)) {
+				curBuf = discardBuf[0:bytesToSkip]
+			}
+			bytesRead, err := reader.Read(curBuf)
+			if err != nil && err != io.EOF {
+				return bytesSkipped, errors.Wrap(err, "Read failed while skipping")
+			}
+			bytesSkipped += int64(bytesRead)
+		}
+	}
+	return bytesSkipped, nil
+}
+func segmentName(baseName string, partNumber int, startOffset int64) string {
+	return baseName + "/" + fmt.Sprintf("%06d-%016d", partNumber, startOffset)
+}
+
+func parseSegmentName(key string) (baseName string, partNumber int, startOffset int64) {
+	lastSep := strings.LastIndexByte(key, '/')
+	if lastSep < 0 {
+		return "", -1, -1
+	}
+	baseName = key[0:lastSep]
+	segmentName := key[lastSep+1 : len(key)]
+	fmt.Sscanf(segmentName, "%d-%d", &partNumber, &startOffset)
+	return
+}
+
+const kMinMultiPartSize = 5 * 1024 * 1024
+const kMaxParts = 10000                 // Maximum number of parts we will use
+const kMaxBufferSize = 10 * 1024 * 1024 // Let's not get too crazy with buffers
+const kSegmentSizeLimit = kMaxParts * kMaxBufferSize
```

Comment:
Is it better to put all constants in a general place? like pkg/util/constants.go in velero plugin

Reply:
Well, this is specific to the S3 plug-in.  I'll make a constants for the plugin